ogr/ogrsf_frmts/gpkg/ogrgeopackagedatasource.cpp

```diff
@@ -8747,3 +8904,323 @@ const GDALRelationship *GDALGeoPackageDataset::GetRelationship( const std::strin
+
+    std::string osRelatedTableType = relationship->GetRelatedTableType();
+    if ( osRelatedTableType.empty() )
+    {
+        osRelatedTableType = "features";
+    }
+
+    // generate mapping table if not set
+    CPLString osMappingTableName = relationship->GetMappingTableName();
+    if ( osMappingTableName.empty() )
+    {
+        int nIndex = 1;
+        osMappingTableName = osLeftTableName + "_" + osRightTableName;
+        while ( FindLayerIndex( osMappingTableName.c_str() ) >= 0 )
+        {
+            nIndex += 1;
+            osMappingTableName.Printf( "%s_%s_%d", osLeftTableName.c_str(), osRightTableName.c_str(), nIndex );
+        }
+
+        // create mapping table
+        // note ideally we'd be setting foreign key constraints on this mapping table too, BUT this will cause sqlite
+        // foreign key mismatches if the base/related table fields aren't primary keys in their tables.
+        // "CONSTRAINT [fk_base_id_%q] FOREIGN KEY (base_id) REFERENCES %q(%q) ON DELETE CASCADE ON UPDATE CASCADE DEFERRABLE INITIALLY DEFERRED, "
+        // "CONSTRAINT [fk_related_id_%q] FOREIGN KEY (related_id) REFERENCES %q(%q) ON DELETE CASCADE ON UPDATE CASCADE DEFERRABLE INITIALLY DEFERRED"
+
+        char* pszSQL = sqlite3_mprintf("CREATE TABLE %q ("
+                                       "id INTEGER PRIMARY KEY AUTOINCREMENT, "
+                                       "base_id INTEGER, "
+                                       "related_id INTEGER"
+                                       ");",
+                                       osMappingTableName.c_str()
+                                       );
+        OGRErr eErr = SQLCommand(hDB, pszSQL);
+        sqlite3_free(pszSQL);
+        if ( eErr != OGRERR_NONE )
+        {
+            failureReason = ( "Could not create mapping table " + osMappingTableName ).c_str();
+            return false;
+        }
+
+        pszSQL = sqlite3_mprintf("CREATE INDEX idx_%q_base_id ON %q (base_id);",
+                                 osMappingTableName.c_str(),
+                                 osMappingTableName.c_str());
+        eErr = SQLCommand(hDB, pszSQL);
+        sqlite3_free(pszSQL);
+        if ( eErr != OGRERR_NONE )
+        {
+            failureReason = ( "Could not create index for " + osMappingTableName + " (base_id)").c_str();
+            return false;
+        }
+
+        pszSQL = sqlite3_mprintf("CREATE INDEX idx_%q_related_id ON %q (related_id);",
+                                 osMappingTableName.c_str(),
+                                 osMappingTableName.c_str());
+        eErr = SQLCommand(hDB, pszSQL);
+        sqlite3_free(pszSQL);
+        if ( eErr != OGRERR_NONE )
+        {
+            failureReason = ( "Could not create index for " + osMappingTableName + " (related_id)").c_str();
+            return false;
+        }
+    }
+    else
+    {
+        const int nMappingTableIndex = FindLayerIndex( osMappingTableName );
```

Comment:
shouldn't nMappingTableIndex be tested against  < 0 ?

Reply:
It's done by the GetLayer call on the next line internally, but I can make it explicit here if you want?

Reply:
Actually I'll replace it with a single call to GetLayerByName, that seems cleanest